cmd/relayer/validate.go

```diff
@@ -0,0 +1,156 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+	rcommon "github.com/athanorlabs/go-relayer/common"
+
+	"github.com/cockroachdb/apd/v3"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+type validator struct {
+	ctx               context.Context
+	ec                *ethclient.Client
+	relayerCommission *apd.Decimal
+	forwarderAddress  ethcommon.Address
+}
+
+func (v *validator) validateTransactionFunc(req *rcommon.SubmitTransactionRequest) error {
+	// validate that:
+	// 1. the `to` address is a swap contract;
+	// 2. the function being called is `claimRelayer`;
+	// 3. the fee passed to `claimRelayer` is equal to or greater
+	// than our desired commission percentage.
+
+	forwarderAddr, err := contracts.CheckSwapFactoryContractCode(
+		v.ctx, v.ec, req.To,
+	)
+	if err != nil {
+		return err
+	}
+
+	if forwarderAddr != v.forwarderAddress {
+		return fmt.Errorf("swap contract does not have expected forwarder address: got %s, expected %s",
+			forwarderAddr,
+			v.forwarderAddress,
+		)
+	}
+
+	// hardcoded, from swap_factory.go bindings
+	claimRelayerSig := ethcommon.FromHex("0x73e4771c")
+	if !bytes.Equal(claimRelayerSig, req.Data[:4]) {
+		return fmt.Errorf("call must be to claimRelayer(); got call to function with sig 0x%x", req.Data[:4])
+	}
+
+	err = validateRelayerFee(req.Data[4:], v.relayerCommission)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateRelayerFee(data []byte, minFeePercentage *apd.Decimal) error {
+	uint256Ty, err := abi.NewType("uint256", "", nil)
+	if err != nil {
+		return fmt.Errorf("failed to create uint256 type: %w", err)
+	}
+
+	bytes32Ty, err := abi.NewType("bytes32", "", nil)
+	if err != nil {
+		return fmt.Errorf("failed to create bytes32 type: %w", err)
+	}
+
+	addressTy, err := abi.NewType("address", "", nil)
+	if err != nil {
+		return fmt.Errorf("failed to create address type: %w", err)
+	}
+
+	arguments := abi.Arguments{
```

Comment:
Is there a contract, with generated code, that we can use to make this easier to cross check and auto-fail if something is incorrect?

Reply:
I tried using the swapABI but it failed to unpack for some reason :/ I slightly updated the code to unpack into a map, it should be clearer now what the args are

Reply:
I can take a stab at using the generated code this evening. If nothing else, it would be educational to understand why it didn't work.

Reply:
Yeah, I couldn't see how to deserialize a slice of 3 differently typed values using the ABI.  My only thought was that maybe we should separate deserialization from validation so we can easily unit test the deserialization.

Reply:
I tried loading the `SwapFactoryABI` into an `abi.ABI` type and doing `Unpack` from there, but it failed, which seems weird considering this method *should* be the same and it worked